src/libfetchers/git-utils.cc

```diff
@@ -909,6 +909,50 @@ struct GitFileSystemObjectSinkImpl : GitFileSystemObjectSink
         addToTree(*pathComponents.rbegin(), oid, GIT_FILEMODE_LINK);
     }
 
+    void createHardlink(const Path & path, const CanonPath & target) override
+    {
+        auto pathComponents = tokenizeString<std::vector<std::string>>(path, "/");
+        if (!prepareDirs(pathComponents, false)) return;
+
+        auto relTarget = CanonPath(path).parent()->makeRelative(target);
+
+        auto dir = pendingDirs.rbegin();
+
+        // For each ../ component at the start, go up one directory.
+        std::string_view relTargetLeft(relTarget);
+        while (hasPrefix(relTargetLeft, "../")) {
```

Comment:
I was trying to see whether all `..`'s are guaranteed to come first when I realised: Is this logic needed? The test file doesn't have any relative paths, suggesting that we could just look up the files directly from the root.

While I like the idea of being lenient towards inputs, we run the risk that a bug here will go unnoticed for a long time, and putting reproducibility at stake.

As far as I can see (I haven't excluded a libarchive quirk), I think we could either
- Figure out a way to make `tar` produce relative hard links (idk if it can)
- Modify the tarball by hand to contain relative paths, see how GNU and BSD tar behave, and use the modified tarball in the test
- ~~Unit testing~~ would potentially test the wrong behavior
- Check that `..` does not occur in any of our path components and `throw UnimplementedError("Nix does not support tarballs with relative hard links yet. Please open an issue.")` if it does. We may never have to implement it.

Reply:
The path returned by libarchive is always absolute (or to be precise, relative to the root of the archive). The relative path here comes from `CanonPath::makeRelative()`, which does indeed always put the `..` elements at the start. The conversion to a relative path is just to deal with the `git_tree_builders` that are in progress (we can't just look up the path from the start of the root, since some parent directories may not have finished yet).